File: Classes/MoPubViewController.m
//
//  Copyright © 2017 SpotX, Inc. All rights reserved.
//

#import "MoPubViewController.h"
#import "InlinePlaybackViewController.h"
#import "ResizablePlaybackViewController.h"
#import "Preferences.h"

@import SpotX;

#define DEFAULT_INTERSTITIAL_ID  @"034709d6e1a4493d8b5da8c7a3e0249c"
#define DEFAULT_REWARDED_ID  @"e3b23c46dd6a41d1a2f049eea9ca2f81"

static NSDictionary* presetUnits(){
  static NSDictionary *units;
  static dispatch_once_t onceToken;
  dispatch_once(&onceToken, ^{
    units = @{
              @"MP4 Ad": @"034709d6e1a4493d8b5da8c7a3e0249c",
              @"VPAID Ad": @"ec9e43b19e8646c19f28f08581153472",
              @"Podded Ad": @"b102fcc5233b48e0b52eaee1fad500c6"
              };
  });
  return units;
}

@interface MoPubViewController ()

@property(nonatomic, strong) SpotXInterstitialAdPlayer* player;

@end

@implementation MoPubViewController {
  __weak IBOutlet UIActivityIndicatorView *_loadingIndicator;
  __weak IBOutlet UIButton *_playInterstitialButton;
  __weak IBOutlet UIButton *_playRewardedButton;
  
  IBOutlet UITextField *_interstitialID;
  IBOutlet UITextField *_rewardedID;
  MPInterstitialAdController *_adController;
  BOOL customAdUnit;
}

- (void)viewDidLoad {
  [super viewDidLoad];
  _interstitialID.text = [Preferences stringForKey:PREF_MOPUB_INTERSTITIAL withDefault:DEFAULT_INTERSTITIAL_ID];
  _rewardedID.text = [Preferences stringForKey:PREF_MOPUB_REWARDED withDefault:DEFAULT_REWARDED_ID];
  
  _interstitialID.delegate = self;
  _rewardedID.delegate = self;
}

- (void)dismissKeyboard {
  [self.view endEditing:YES];
  [_interstitialID resignFirstResponder];
  [_rewardedID resignFirstResponder];
}

- (NSString *)interstitialID {
  NSString *adId = _interstitialID.text;
  if (![adId length]) {
    adId = DEFAULT_INTERSTITIAL_ID;
    _interstitialID.text = adId;
  }
  return adId;
}

- (NSString *)rewardedID {
  NSString *adId = _rewardedID.text;
  if (![adId length]) {
    adId = DEFAULT_REWARDED_ID;
    _rewardedID.text = adId;
  }
  return adId;
}

- (void)updateAdIDs {
  [Preferences setString:[self interstitialID] forKey:PREF_MOPUB_INTERSTITIAL];
  [Preferences setString:[self rewardedID] forKey:PREF_MOPUB_REWARDED];
}

- (IBAction)playInterstitial:(id)sender {
  [self dismissKeyboard];
  _playInterstitialButton.enabled = _playRewardedButton.enabled = NO;
  
  // MoPub requires a global API key as it can't be set per-request
  [SpotX setAPIKey:SPOTX_API_KEY];
  
  _adController = [MPInterstitialAdController interstitialAdControllerForAdUnitId:[self interstitialID]];
  if(_adController != nil){
    _adController.delegate = self;
    [_loadingIndicator startAnimating];
    [_adController loadAd];
  } else {
    [self showMessage:@"Failed to create MoPub ad controller"];
  }
}

- (IBAction)playRewarded:(id)sender {
  [self dismissKeyboard];
  _playInterstitialButton.enabled = _playRewardedButton.enabled = NO;
  
  // MoPub requires a global API key as it can't be set per-request
  [SpotX setAPIKey:SPOTX_API_KEY];
  
  [[MoPub sharedInstance] initializeRewardedVideoWithGlobalMediationSettings:nil delegate:self];
  [MPRewardedVideo loadRewardedVideoAdWithAdUnitID:[self rewardedID] withMediationSettings:nil];
  [_loadingIndicator startAnimating];
}

- (void)showAdFailure {
  [_loadingIndicator stopAnimating];
  _adController = nil;
  _playInterstitialButton.enabled = _playRewardedButton.enabled = YES;
  [self showMessage:@"Ad playback failed"];
}

#pragma mark - Pop-up ad unit selector

-(void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
  if ([segue.identifier isEqualToString:@"pickerSegue"]) {
    // Showing the pop-up with picker view
    UIViewController *vc = [segue destinationViewController];
    vc.modalPresentationStyle = UIModalPresentationPopover;
    vc.popoverPresentationController.delegate = self;
    
    // Set ourselves as the picker source
    UIPickerView* picker = vc.view.subviews[0];
    picker.delegate = self;
    picker.dataSource = self;
    // Select the current ad unit (if any)
    NSUInteger adUnitIndex = [presetUnits().allValues indexOfObject:[self interstitialID]];
    if(adUnitIndex != NSNotFound){
      customAdUnit = NO;
      [picker selectRow:adUnitIndex inComponent:0 animated:NO];
    } else {
      customAdUnit = YES;
      [picker selectRow:presetUnits().count inComponent:0 animated:NO];
    }
  }
}

-(UIModalPresentationStyle)adaptivePresentationStyleForPresentationController:(UIPresentationController *)controller {
  return UIModalPresentationNone;
}

- (UIModalPresentationStyle)adaptivePresentationStyleForPresentationController:(UIPresentationController *)controller traitCollection:(UITraitCollection *)traitCollection {
  return UIModalPresentationNone;
}

- (NSInteger)numberOfComponentsInPickerView:(UIPickerView *)pickerView {
  return 1;
}

- (NSInteger)pickerView:(UIPickerView *)pickerView numberOfRowsInComponent:(NSInteger)component {
  return presetUnits().count + (customAdUnit ? 1 : 0);
}

- (NSString *)pickerView:(UIPickerView *)pickerView titleForRow:(NSInteger)row forComponent:(NSInteger)component {
  if(row >= presetUnits().count){
    return @"---";
  }
  return presetUnits().allKeys[row];
}

- (void)pickerView:(UIPickerView *)pickerView didSelectRow:(NSInteger)row inComponent:(NSInteger)component {
  // Set Ad Unit text field to the selected ad
  NSString* adUnit = presetUnits().allValues[row];
  [_interstitialID setText:adUnit];
  [self updateAdIDs];
}

#pragma mark - UITextFieldDelegate

- (void)textFieldDidEndEditing:(UITextField *)textField {
  [self updateAdIDs];
}

- (BOOL)textFieldShouldReturn:(UITextField *)textField {
  [textField resignFirstResponder];
  return YES;
}

#pragma mark - MPInterstitialAdControllerDelegate

- (void)interstitialDidLoadAd:(MPInterstitialAdController *)interstitial
{
  NSLog(@"MoPub:interstitialDidLoadAd");
  // Ad loaded; play the ad
  [_loadingIndicator stopAnimating];
  [_adController showFromViewController:self];
}

- (void)interstitialDidFailToLoadAd:(MPInterstitialAdController *)interstitial
{
  NSLog(@"MoPub:interstitialDidFailToLoadAd");
  [self showAdFailure];
}

- (void)interstitialDidDisappear:(MPInterstitialAdController *)interstitial
{
  NSLog(@"MoPub:interstitialDidDisappear");
  _adController = nil;
  _playInterstitialButton.enabled = _playRewardedButton.enabled = YES;
}

- (void)interstitialDidExpire:(MPInterstitialAdController *)interstitial
{
  NSLog(@"MoPub:interstitialDidExpire");
  _adController = nil;
  _playInterstitialButton.enabled = _playRewardedButton.enabled = YES;
}

- (void)interstitialDidReceiveTapEvent:(MPInterstitialAdController *)interstitial
{
  // Use this event to monitor click-thru
  NSLog(@"MoPub:interstitialDidReceiveTapEvent");
}

#pragma mark - MPRewardedVideoDelegate

- (void)rewardedVideoAdDidLoadForAdUnitID:(NSString *)adUnitID {
  NSLog(@"MoPub:rewardedVideoAdDidLoad");
  [_loadingIndicator stopAnimating];
  [MPRewardedVideo presentRewardedVideoAdForAdUnitID:adUnitID fromViewController:self withReward:nil];    // add your reward here
}

- (void)rewardedVideoAdDidFailToLoadForAdUnitID:(NSString *)adUnitID error:(NSError *)error {
  NSLog(@"MoPub:rewardedVideoAdDidFailToLoad");
  [self showAdFailure];
}

- (void)rewardedVideoAdDidFailToPlayForAdUnitID:(NSString *)adUnitID error:(NSError *)error {
  NSLog(@"MoPub:rewardedVideoAdDidFailToPlay");
  [self showAdFailure];
}

- (void)rewardedVideoAdDidDisappearForAdUnitID:(NSString *)adUnitID {
  NSLog(@"MoPub:rewardedVideoAdDidDisappear");
  _playInterstitialButton.enabled = _playRewardedButton.enabled = YES;
}

- (void)rewardedVideoAdDidExpireForAdUnitID:(NSString *)adUnitID {
  NSLog(@"MoPub:rewardedVideoAdDidExpire");
  _playInterstitialButton.enabled = _playRewardedButton.enabled = YES;
}

@end

File: Classes/MoPubViewController.swift
//
//  Copyright © 2018 SpotX. All rights reserved.
//

import UIKit
import SpotX

let DEFAULT_INTERSTITIAL_ID = "034709d6e1a4493d8b5da8c7a3e0249c"
let DEFAULT_REWARDED_ID     = "e3b23c46dd6a41d1a2f049eea9ca2f81"

let DEFAULT_AD_UNITS = [
  "MP4 Ad":     "034709d6e1a4493d8b5da8c7a3e0249c",
  "VPAID Ad":   "ec9e43b19e8646c19f28f08581153472",
  "Podded Ad":  "b102fcc5233b48e0b52eaee1fad500c6"
]

class MoPubViewController: ViewControllerBase, MPInterstitialAdControllerDelegate, MPRewardedVideoDelegate, UITextFieldDelegate, UIPopoverPresentationControllerDelegate, UIPickerViewDataSource, UIPickerViewDelegate {
  @IBOutlet var _loadingIndicator: UIActivityIndicatorView!
  @IBOutlet var _playInterstitialButton: UIButton!
  @IBOutlet var _playRewardedButton: UIButton!
  @IBOutlet var _interstitialID: UITextField!
  @IBOutlet var _rewardedID: UITextField!
  
  var _adController: MPInterstitialAdController?
  
  var customAdUnit = false      // for pop-up control
  
  override func viewDidLoad() {
    super.viewDidLoad()
    _interstitialID.text = Preferences.string(forKey: .MOPUB_INTERSTITIAL) ?? DEFAULT_INTERSTITIAL_ID
    _rewardedID.text = Preferences.string(forKey: .MOPUB_REWARDED) ?? DEFAULT_REWARDED_ID
    
    _interstitialID.delegate = self
    _rewardedID.delegate = self
  }
  
  override func dismissKeyboard() {
    self.view.endEditing(true)
    _interstitialID.resignFirstResponder()
    _rewardedID.resignFirstResponder()
  }
  
  @IBAction func playInterstitial(_ sender: Any) {
    dismissKeyboard()
    _playInterstitialButton.isEnabled = false
    _playRewardedButton.isEnabled = false
    
    // MoPub requires a global API key as it can't be set per-request
    SpotX.setAPIKey(Preferences.SPOTX_API_KEY)
    
    _adController = MPInterstitialAdController(forAdUnitId: interstitialID())
    guard let adController = _adController else {
      showMessage("Failed to create MoPub ad controller")
      return
    }
    adController.delegate = self
    _loadingIndicator.startAnimating()
    adController.loadAd()
  }
  
  @IBAction func playRewarded(_ sender: Any) {
    dismissKeyboard()
    _playInterstitialButton.isEnabled = false
    _playRewardedButton.isEnabled = false
    
    // MoPub requires a global API key as it can't be set per-request
    SpotX.setAPIKey(Preferences.SPOTX_API_KEY)
    
    MoPub.sharedInstance().initializeRewardedVideo(withGlobalMediationSettings: [Any](), delegate: self)
    MPRewardedVideo.loadAd(withAdUnitID: rewardedID(), withMediationSettings: [Any]())
    _loadingIndicator.startAnimating()
  }
  
  func interstitialID() -> String {
    let adId = _interstitialID.text ?? ""
    if adId == "" {
      _interstitialID.text = DEFAULT_INTERSTITIAL_ID
      return DEFAULT_INTERSTITIAL_ID
    }
    return adId
  }
  
  func rewardedID() -> String {
    let adId = _rewardedID.text ?? ""
    if adId == "" {
      _rewardedID.text = DEFAULT_REWARDED_ID
      return DEFAULT_REWARDED_ID
    }
    return adId
  }
  
  func updateAdIDs() {
    Preferences.set(interstitialID(), forKey: .MOPUB_INTERSTITIAL)
    Preferences.set(rewardedID(), forKey: .MOPUB_REWARDED)
  }
  
  func showAdFailure() {
    _loadingIndicator.stopAnimating()
    _adController = nil
    _playInterstitialButton.isEnabled = true
    _playRewardedButton.isEnabled = true
    showMessage("Ad playback failed")
  }
  
  // MARK: Pop-up ad unit selector
  
  override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
    if segue.identifier == "pickerSegue" {
      // Showing the pop-up with picker view
      let vc = segue.destination
      vc.modalPresentationStyle = .popover
      vc.popoverPresentationController?.delegate = self
      
      let picker = vc.view.subviews.first as! UIPickerView
      picker.delegate = self
      picker.dataSource = self
      let adUnitArray = Array(DEFAULT_AD_UNITS.values)
      if let adUnitIndex = adUnitArray.index(of: interstitialID()) {
        customAdUnit = false
        picker.selectRow(adUnitIndex.hashValue, inComponent: 0, animated: false)
      } else {
        customAdUnit = true
        picker.selectRow(DEFAULT_AD_UNITS.count, inComponent: 0, animated: false)
      }
    }
  }
  
  func adaptivePresentationStyle(for controller: UIPresentationController) -> UIModalPresentationStyle {
    return .none
  }
  
  func adaptivePresentationStyle(for controller: UIPresentationController, traitCollection: UITraitCollection) -> UIModalPresentationStyle {
    return .none
  }
  
  func numberOfComponents(in pickerView: UIPickerView) -> Int {
    return 1
  }
  
  func pickerView(_ pickerView: UIPickerView, numberOfRowsInComponent component: Int) -> Int {
    return DEFAULT_AD_UNITS.count + (customAdUnit ? 1 : 0)
  }
  
  func pickerView(_ pickerView: UIPickerView, titleForRow row: Int, forComponent component: Int) -> String? {
    if row >= DEFAULT_AD_UNITS.count {
      return "---"
    }
    return Array(DEFAULT_AD_UNITS.keys)[row]
  }
  
  func pickerView(_ pickerView: UIPickerView, didSelectRow row: Int, inComponent component: Int) {
    let adUnit = Array(DEFAULT_AD_UNITS.values)[row]
    _interstitialID.text = adUnit
    updateAdIDs()
  }
  
  // MARK: UITextFieldDelegate
  
  func textFieldShouldReturn(_ textField: UITextField) -> Bool {
    textField.resignFirstResponder()
    return true
  }
  
  func textFieldDidEndEditing(_ textField: UITextField) {
    updateAdIDs()
  }
  
  // MARK: MPInterstitialAdControllerDelegate
  
  func interstitialDidLoadAd(_ interstitial: MPInterstitialAdController!) {
    NSLog("MoPub:interstitialDidLoadAd")
    // Ad loaded; play the ad
    _loadingIndicator.stopAnimating()
    _adController!.show(from: self)
  }
  
  func interstitialDidFail(toLoadAd interstitial: MPInterstitialAdController!) {
    NSLog("MoPub:interstitialDidFail")
    showAdFailure()
  }
  
  func interstitialDidDisappear(_ interstitial: MPInterstitialAdController!) {
    NSLog("MoPub:interstitialDidDisappear")
    _adController = nil
    _playInterstitialButton.isEnabled = true
    _playRewardedButton.isEnabled = true
  }
  
  func interstitialDidExpire(_ interstitial: MPInterstitialAdController!) {
    NSLog("MoPub:interstitialDidExpire")
    _adController = nil
    _playInterstitialButton.isEnabled = true
    _playRewardedButton.isEnabled = true
  }
  
  func interstitialDidReceiveTapEvent(_ interstitial: MPInterstitialAdController!) {
    // Use this event to monitor click-thru
    NSLog("MoPub:interstitialDidReceiveTapEvent")
  }
  
  // MARK: MPRewardedVideoDelegate
  
  func rewardedVideoAdDidLoad(forAdUnitID adUnitID: String!) {
    NSLog("MoPub:rewardedVideoAdDidLoad")
    _loadingIndicator.stopAnimating()
    MPRewardedVideo.presentAd(forAdUnitID: adUnitID, from: self, with: nil)   // add your reward here
  }
  
  func rewardedVideoAdDidFailToLoad(forAdUnitID adUnitID: String!, error: Error!) {
    NSLog("MoPub:rewardedVideoAdDidFailToLoad")
    showAdFailure()
  }
  
  func rewardedVideoAdDidFailToPlay(forAdUnitID adUnitID: String!, error: Error!) {
    NSLog("MoPub:rewardedVideoAdDidFailToPlay")
    showAdFailure()
  }
  
  func rewardedVideoAdDidReceiveTapEvent(forAdUnitID adUnitID: String!) {
    // Use this event to monitor click-thru
    NSLog("MoPub:rewardedVideoAdDidReceiveTapEvent")
  }
  
  func rewardedVideoAdDidDisappear(forAdUnitID adUnitID: String!) {
    NSLog("MoPub:rewardedVideoAdDidDisappear")
    _playInterstitialButton.isEnabled = true
    _playRewardedButton.isEnabled = true
  }
  
  func rewardedVideoAdDidExpire(forAdUnitID adUnitID: String!) {
    NSLog("MoPub:rewardedVideoAdDidExpire")
    _playInterstitialButton.isEnabled = true
    _playRewardedButton.isEnabled = true
  }
  
}
